components/automate-deployment/pkg/backup/artifact_streams.go

```diff
@@ -0,0 +1,449 @@
+package backup
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"go.uber.org/multierr"
+)
+
+// ArtifactStream streams a list of artifacts keys
+// by name. A stream may not produce the same value
+// twice.
+// Returns io.EOF when no more elements will be produced
+type ArtifactStream interface {
+	Next() (string, error)
+	Close() error
+}
+
+// ErrStream is an ArtifactStream that always returns
+// the given error
+func ErrStream(err error) ArtifactStream {
+	return &errStream{
+		err: err,
+	}
+}
+
+type errStream struct {
+	err error
+}
+
+func (e errStream) Next() (string, error) {
+	return "", e.err
+}
+
+func (e errStream) Close() error {
+	return nil
+}
+
+// EmptyStream is an ArtifactStream that always returns
+// io.EOF
+func EmptyStream() ArtifactStream {
+	return ErrStream(io.EOF)
+}
+
+// PeekableArtifactStream is an artifact stream that
+// allows peeking at the next element without consuming
+// it
+type PeekableArtifactStream interface {
+	ArtifactStream
+	// Peek returns the next element that will be returned
+	// when Next is called. It can be called multiple times
+	// and will keep generating the same value until Next
+	// is called
+	Peek() (string, error)
+}
+
+type peekableArtifactStream struct {
+	stream ArtifactStream
+	next   *string
```

Comment:
Are we using `*string` here primarily to differentiate between a valid `""` value and a non-existent value or is there an additional reason?

Reply:
i didn't want to think about if an empty string was valid or non-existent. But i think we could say empty string is the same as non-existent as other iterators filter them out